airflow_prometheus_exporter/prometheus_exporter.py

```diff
@@ -546,50 +609,54 @@ def get_sla_miss_tasks():
                 DelayAlertMetaData.group_business_line,
                 DelayAlertMetaData.group_pagerduty,
                 DelayAlertMetaData.inhibit_rule,
-                DelayAlertMetaData.latest_successful_run,
                 DelayAlertMetaData.link,
                 DelayAlertMetaData.severity,
                 DelayAlertMetaData.sla_interval,
                 DelayAlertMetaData.sla_time,
                 DelayAlertMetaData.task_id,
-                execution_dt_query.c.execution_date,
-                execution_dt_query.c.schedule_interval,
-                execution_dt_query.c.state,
+                LatestSuccessfulRun.execution_date.label("latest_successful_run"),
             )
             .join(
-                execution_dt_query,
+                active_task_query,
                 and_(
-                    execution_dt_query.c.dag_id == DelayAlertMetaData.dag_id,
-                    execution_dt_query.c.task_id == DelayAlertMetaData.task_id,
+                    DelayAlertMetaData.dag_id == active_task_query.c.dag_id,
+                    DelayAlertMetaData.task_id == active_task_query.c.task_id,
+                )
+            )
+            .join(
+                LatestSuccessfulRun,
+                and_(
+                    DelayAlertMetaData.dag_id == LatestSuccessfulRun.dag_id,
+                    DelayAlertMetaData.task_id == LatestSuccessfulRun.task_id,
                 ),
+                isouter=True
+            )
+            .filter(
+                DelayAlertMetaData.task_id.isnot(None),
             )
-            .filter(DelayAlertMetaData.sla_interval.isnot(None))
             .all()
         )
-        execution_dates = {}
-        max_execution_dates = {}
-        for run in runs:
-            key = (run.dag_id, run.task_id)
-            if key not in execution_dates:
-                execution_dates[key] = []
-            if key not in max_execution_dates and run.state == "success":
-                max_execution_dates[key] = run.execution_date
-            execution_dates[key].append(
-                {
-                    "execution_date": run.execution_date,
-                    "state": run.state,
-                }
-            )
-
-        sla_miss_metrics = []
-        processed_runs = set()
+
+        metrics = []
+        insert_dict = {}
+        update_dict = {}
         for run in runs:
             key = (run.dag_id, run.task_id)
-            if key in processed_runs or key not in max_execution_dates:
+
+            max_execution_date = max_execution_dates.get(
+                key,
+                datetime.datetime(2020, 1, 1, tzinfo=datetime.timezone.utc)
+            )
+            if max_execution_date is None:
```

Comment:
will max_execution_date ever be None?
since you are doing a max_execution_dates.get() with the default set to `datetime.datetime(2020, 1, 1, tzinfo=datetime.timezone.utc)`

Reply:
kind of seems like the logic here should be

```
            max_execution_date = max_execution_dates.get(
                key,
                run.latest_successful_run
            )
            if max_execution_date is None:
                continue
            elif run.latest_successful_run is None and max_execution_date:
                insert_dict[(run.dag_id, run.task_id)] = max_execution_date
            elif max_execution_date > run.latest_successful_run:
                update_dict[(run.dag_id, run.task_id)] = max_execution_date
```

Reply:
The `max_execution_date` here before the if conditions is from `execution_dates`. The execution_dates may have state that was `success` but `running` now due to re-processing. We use `latest_successful_run` to store the max successful execution date we ever seen. 

The `max_execution_date` maybe `None` because of a dag is new and never had any successful run. Here I set the default date time to `2020-01-01` to have the lowest date outside of our retention window. 

This logic enables us to have newly setup alert to trigger sla_miss if it never has a successful run.

That's why we can't have `max_execution_date` as the default value of `run.latest_successful_run`.